cupyx/scipy/sparse/compressed.py

```diff
@@ -33,6 +33,51 @@ class _compressed_sparse_matrix(sparse_data._data_matrix,
         ''',
         'compress_getitem_complex')
 
+    _compress_setitem_kern = core.ElementwiseKernel(
+        'T value, S ind, int32 index', 'raw T d',
+        'atomicExch(&d[index], value);',
+        'compress_setitem',
+        preamble='''
+            #if __CUDA_ARCH__ < 600
+            __device__ double atomicExch(double* address, double val)
+            {
+                unsigned long long int* address_as_ull =
+                                          (unsigned long long int*)address;
+                unsigned long long int old = *address_as_ull, compare;
+                do {
+                    compare = old;
+                    old = atomicCAS(address_as_ull, compare,
+                                            __double_as_longlong(val));
+                } while (compare != old);
+                return __longlong_as_double(old);
+            }
+            #endif
+            ''')
+    _compress_setitem_complex_kern = core.ElementwiseKernel(
+        'T value_real, T value_imag, S ind, int32 index',
+        'raw T real, raw T imag',
+        '''
+          atomicExch(&real[index], value_real);
+          atomicExch(&imag[index], value_imag);
+        ''',
+        'compress_setitem_complex',
+        preamble='''
+            #if __CUDA_ARCH__ < 600
```

Comment:
This code is the same as above, store it in a `setitem_preamble` variable.

Reply:
Also add a comment with the reference where you get the code from
`https://www.micc.unifi.it/bertini/download/gpu-programming-basics/2017/gpu_cuda_5.pdf`
(I don't know if this is the original source ...)